### PR TITLE
Set default rootless sigstore

### PR DESCRIPTION
**Pull request by QiWang19**
Set default rootless sigstore to ~/.local/share/containers/sigstore if the caller is non-root.
Fix rootless podman push --sign-by.

Signed-off-by: Qi Wang <qiwan@redhat.com>

### CODE REVIEW

**Comment by QiWang19** [comment]
@mtrmac PTAL

**Comment by QiWang19** [comment]
Add `GetRootlessUID()` to internal subpackage according to https://github.com/containers/image/pull/1027#issuecomment-679320458, can use unshare.GetRootlessUID() from containers/storage here?

File: docker/lookaside.go
Comment (mtrmac) [inline docker/lookaside.go:190]: It would be rather surprising if the value visible in the config file (and only one specific value) were silently overridden. The conversation in #1027 suggested that:
- We would ship config files without a default (or no config file at all); but users upgrading from an older release could/would retain the old configuration
- The code would contain a built-in default that only applied if the config files did not specify anything.

Also, the man page should be updated.

File: docker/lookaside.go
Comment (mtrmac) [inline docker/lookaside.go:35]: `User`, `Nonroot`, `Unprivileged`, not `Rootless`, please; that means something more specific.

File: internal/uid/uid.go
Comment (mtrmac) [inline internal/uid/uid.go:1]: (Maybe `rootless`? Then this could be `rootless.GetEuid`.)

File: internal/uid/uid.go
Comment (mtrmac) [inline internal/uid/uid.go:9]: Please call this from `pkg/blobinfocache/default.go`, and drop the copy in that file.

File: internal/uid/uid.go
Comment (mtrmac) [inline internal/uid/uid.go:8]:
```suggestion
// GetRootlessUID returns the UID of the current user (in the parent userNS, if any)
//
// Podman and similar software, in “rootless” configuration, when run as a non-root
// user, very early switches to a user namespace, where Geteuid() == 0 (but does not
// switch to a limited mount namespace); so, code relying on Geteuid() would use 
// system-wide paths in e.g. /var, when the user is actually not privileged to write to
// them, and expects state to be stored in the home directory.
//
// If Podman is setting up such a user namespace, it records the original UID in an
// environment variable, allowing us to make choices based on the actual user’s identity.
```
or something like that (if the above is accurate), to give newcomers more context.

File: internal/uid/uid.go
Comment (mtrmac) [inline internal/uid/uid.go:9]: This should all talk about EUID, in all of function/variable names and comments, because that’s what the value apparently is (except the environment variable must not be renamed).

File: docker/lookaside.go
Comment (QiWang19) [reply docker/lookaside.go]: @mtrmac  PTAL.  Use the old configuration if the default configuration exists. And when euid not 0, only use the built-in default if there's no per-user configuration.

**Review by mtrmac — COMMENTED** [review]
Oh… the rest of the client will now need to account for the state that sigstore is no longer “opt-in”. Especially after the `/var` path is added, the various code paths that assume `signatureStorageBase` can be `nil` become dead code.
- As the rest of the code is now, X-R-S-S would never be used, because sigstore is preferred; so we need to either prefer X-R-S-S (#385), or add a new configuration semantics to explicitly say “don’t use sigstore, use X-R-S-S or nothing”. I guess we should just do #385 ; the likelihood that a registry suddenly gains X-R-S-S is getting smaller every month.

**Comment by QiWang19** [comment]
@mtrmac PTAL. I added the changes from #385. But the CI is not happy, will check the test failures.

File: docker/lookaside.go
Comment (mtrmac) [inline docker/lookaside.go:37]:
```suggestion
// defaultDockerDir is the default sigstore directory for root
```

File: docs/containers-registries.d.5.md
Comment (mtrmac) [inline docs/containers-registries.d.5.md:81]: That’s not the actual condition.

File: docs/containers-registries.d.5.md
Comment (mtrmac) [inline docs/containers-registries.d.5.md:82]: All of this should be in the ”Registries, Scopes and Search Order” section, maybe in an extra named subsection.

File: docs/containers-registries.d.5.md
Comment (mtrmac) [inline docs/containers-registries.d.5.md:84]: Consider adding an extra “example” for no configuration.

File: docker/lookaside.go
Comment (mtrmac) [inline docker/lookaside.go:196]:
```suggestion
	logrus.Debugf(" No signature storage configuration found for %s, using default %s", ref.PolicyConfigurationIdentity(), u.String())
```
(in both cases)

File: docker/lookaside.go
Comment (mtrmac) [inline docker/lookaside.go:188]: > `// config.signatureTopLevel returns an URL string configured in config for ref, for write access if “write”.`

i.e. this could return `nil` if `config` does not specify a configuration.

Moving these defaults up to `configuredSignatureStorageBase` would also avoid the unnecessary roundtrip between `url.URL`, a `string`, and back.

--- 

Either way the documentation of the function needs updating, if only to update the “if no signature storage should be used.” part.

File: docker/lookaside_test.go
Comment (mtrmac) [inline docker/lookaside_test.go:43]: No, please test both: split `configuredSignatureStorageBaseForEUID` from `configuredSignatureStorageBase`, document that it only exists for tests, and test the `…ForEUID` variant here, for both the root and non-root cases. Or maybe split a `builtinDefaultSignatureStorageDir` with an EUID parameter, test both only for that one small helper, and here assume that tests are not running as root.

File: docker/lookaside_test.go
Comment (mtrmac) [inline docker/lookaside_test.go:274]: Moving the default directories from `signatureTopLevel` to `configuredSignatureStorageBase` would also allow not changing these tests.

File: docker/lookaside.go
Comment (mtrmac) [inline docker/lookaside.go:197]: After all this settles and tests work, the code and documentation that assumes `configuredSignatureStorageBase` can be `nil` should be removed/updated.

**Comment by mtrmac** [comment]
> I added the changes from #385. But the CI is not happy, will check the test failures.

I think that’s because this adds only half of #385; the source implementation needs the change as well, or it will try reading the sigstore instead of X-R-S-S.

File: docs/containers-registries.d.5.md
Comment (mtrmac) [inline docs/containers-registries.d.5.md:64]: Maybe
```suggestion
### Built-in Defaults
```
everything is about signature storage in this file.

File: docs/containers-registries.d.5.md
Comment (mtrmac) [inline docs/containers-registries.d.5.md:67]: This should say how the value is chosen.

**Review by mtrmac — COMMENTED** [review]
ACK overall.

Would it be useful to make some variant of configuredSignatureStorageBase public for Podman? That’s up mostly to you, I think.

File: docker/lookaside.go
Comment (mtrmac) [inline docker/lookaside.go:56]: This does not work for external callers if we don’t export `signatureStorageURL` (can the result even be used in Podman without an explicit `*url.URL` cast?)

… but there’s not _that much_ benefit to doing just that, because any caller would hard-code the semantics of the signature storage (in particular the sequential indices) anyway. We would need a very generic ReadSignatures/WriteSignatures/DeleteSignature API, which seems more effort than is strictly necessary right now.

So…

File: docker/lookaside.go
Comment (mtrmac) [inline docker/lookaside.go:59]: … maybe name this something like `SignatureStorageBaseURL` (`RegistriesDSignatureStorageBaseURL`? `RegistriesDSigstoreBaseURL`?) and have it return an URL directly?

The documentation should also link to the “docker/distribution registries—separate storage” section of `docs/signature-protocols.md` for explanation of what a “base URL” is, and explicitly say something like
> Warning:  This function only exposes configuration in registries.d; just because this function returns an URL does not mean that the URL will be used by c/image/docker (e.g. if the registry natively supports X-R-S-S).

File: docker/lookaside.go
Comment (mtrmac) [inline docker/lookaside.go:70]: (Absolutely non-blocking: It would be a tiny bit more readable to start with `topLevel != ""`, so that the uses of `topLevel` are close together.)

File: docker/lookaside.go
Comment (mtrmac) [inline docker/lookaside.go:56]: The `nil` value can not happen any longer.

File: docker/lookaside_test.go
Comment (mtrmac) [inline docker/lookaside_test.go:321]: Are the `*url.URL` casts truly necessary here?

File: docker/docker_image_dest.go
Comment (mtrmac) [inline docker/docker_image_dest.go:85]: This is now unreachable; so this function can just be `return nil`, or the error returned might be “Internal error”+description of the inconsistency.

File: docker/docker_image_dest.go
Comment (mtrmac) [inline docker/docker_image_dest.go:486]: This is now unreachable; consistently with `SupportsSignatures`, either assume `d.c.signatureBase` is non-`nil`, or report an internal error.

File: docker/docker_image_src.go
Comment (mtrmac) [inline docker/docker_image_src.go:304]: This is now unreachable, handle it consistently with `docker_image_dest`.

**Comment by QiWang19** [comment]
@mtrmac PTAL

File: docker/docker_image_src.go
Comment (mtrmac) [inline docker/docker_image_src.go:477]: OK; please document for `newDockerClientFromRef` that in the return value, `signatureBase` is always set.

File: docker/lookaside.go
Comment (mtrmac) [inline docker/lookaside.go:218]: The checks for the return value being `nil` in callers can be removed as well.

File: docker/lookaside.go
Comment (QiWang19) [reply docker/lookaside.go:56]: the `dockerReference` also needs to be exported

Comment (mtrmac) [reply docker/lookaside.go:56]: Exporting that would not directly help; users of `alltransports.ParseImageName` or `docker.NewReference` only get a generic `types.ImageReference` interface.

`GetRepositoryTags` already handles that by accepting a `types.ImageReference` and making a runtime check. That’s ugly but convenient enough for callers, and impossible to misuse.

Comment (QiWang19) [reply docker/lookaside.go]: Thanks.  Add the same type assertion check inside `SignatureStorageBaseURL`

File: docker/docker_image_dest.go
Comment (QiWang19) [reply docker/docker_image_dest.go:85]: I keep them as internal errors.

**Comment by QiWang19** [comment]
@mtrmac Can I get a review before you leave for PTO?

File: docs/containers-registries.d.5.md
Comment (mtrmac) [inline docs/containers-registries.d.5.md:66]: (Maybe something like
```suggestion
If no `docker` section can be found for the container image, and no `default-docker` section is configured
```
to go from the most-specific to the most general, in the “order of evaluation”)

**Review by mtrmac — COMMENTED** [review]
ACK, please rebase on top of the recent typo fix.

(And ideally make sure Podman can benefit from `SignatureStorageBaseURL`.)

**Comment by mtrmac** [comment]
LGTM after dropping the  `docker/docker_transport.go` part, actually.

**Review by mtrmac — APPROVED** [review]
Thanks!

**Pull request merged**
